okio/src/jvmMain/kotlin/okio/AsyncTimeout.kt

```diff
@@ -250,77 +288,67 @@ open class AsyncTimeout : Timeout() {
+      if (head == null) {
+        head = AsyncTimeout()
+        Watchdog().start()
+      }
 
-        val now = System.nanoTime()
-        if (timeoutNanos != 0L && hasDeadline) {
-          // Compute the earliest event; either timeout or deadline. Because nanoTime can wrap
-          // around, minOf() is undefined for absolute values, but meaningful for relative ones.
-          node.timeoutAt = now + minOf(timeoutNanos, node.deadlineNanoTime() - now)
-        } else if (timeoutNanos != 0L) {
-          node.timeoutAt = now + timeoutNanos
-        } else if (hasDeadline) {
-          node.timeoutAt = node.deadlineNanoTime()
-        } else {
-          throw AssertionError()
-        }
+      val now = System.nanoTime()
+      if (timeoutNanos != 0L && hasDeadline) {
+        // Compute the earliest event; either timeout or deadline. Because nanoTime can wrap
+        // around, minOf() is undefined for absolute values, but meaningful for relative ones.
+        node.timeoutAt = now + minOf(timeoutNanos, node.deadlineNanoTime() - now)
+      } else if (timeoutNanos != 0L) {
+        node.timeoutAt = now + timeoutNanos
+      } else if (hasDeadline) {
+        node.timeoutAt = node.deadlineNanoTime()
+      } else {
+        throw AssertionError()
+      }
 
-        // Insert the node in sorted order.
-        val remainingNanos = node.remainingNanos(now)
-        var prev = head!!
-        while (true) {
-          if (prev.next == null || remainingNanos < prev.next!!.remainingNanos(now)) {
-            node.next = prev.next
-            prev.next = node
-            if (prev === head) {
-              // Wake up the watchdog when inserting at the front.
-              condition.signal()
-            }
-            break
+      // Insert the node in sorted order.
+      val remainingNanos = node.remainingNanos(now)
+      var prev = head!!
+      while (true) {
+        if (prev.next == null || remainingNanos < prev.next!!.remainingNanos(now)) {
+          node.next = prev.next
+          prev.next = node
+          if (prev === head) {
+            // Wake up the watchdog when inserting at the front.
+            condition.signal()
           }
-          prev = prev.next!!
+          break
         }
+        prev = prev.next!!
       }
     }
 
     /** Returns true if the timeout occurred. */
-    private fun cancelScheduledTimeout(node: AsyncTimeout): Boolean {
-      if (!node.inQueue) return false
-      node.inQueue = false
-
-      // Remove the node from the linked list.
+    private fun removeFromQueue(node: AsyncTimeout) {
       var prev = head
       while (prev != null) {
         if (prev.next === node) {
           prev.next = node.next
           node.next = null
-          return false
+          return
         }
         prev = prev.next
       }
 
-      // The node wasn't found in the linked list: it must have timed out!
-      return true
```

Comment:
Now that we keep a state machine we don’t need to traverse the linked list to learn this